protocol/datasync/transport.go

```diff
@@ -39,21 +50,100 @@ func (t *NodeTransport) Watch() transport.Packet {
 }
 
 func (t *NodeTransport) Send(_ state.PeerID, peer state.PeerID, payload protobuf.Payload) error {
+	var lastError error
 	if t.dispatch == nil {
 		return errNotInitialized
 	}
 
-	data, err := proto.Marshal(&payload)
-	if err != nil {
-		return err
+	payloads := splitPayloadInBatches(&payload, int(t.maxMessageSize))
+	for _, payload := range payloads {
+
+		data, err := proto.Marshal(payload)
+		if err != nil {
+			return err
+		}
+
+		publicKey, err := datasyncpeer.IDToPublicKey(peer)
+		if err != nil {
+			return err
+		}
+		err = t.dispatch(context.Background(), publicKey, data, payload)
+		if err != nil {
+			lastError = err
+			t.logger.Error("failed to send message", zap.Error(err))
+			continue
+		}
 	}
+	return lastError
+}
+
+func splitPayloadInBatches(payload *protobuf.Payload, maxSizeBytes int) []*protobuf.Payload {
+	newPayload := &protobuf.Payload{}
+	var response []*protobuf.Payload
+	currentSize := payloadTagSize
 
-	publicKey, err := datasyncpeer.IDToPublicKey(peer)
-	if err != nil {
-		return err
+	// this is not going to be 100% accurate, but should be fine in most cases, faster
+	// than using proto.Size
+	for _, ack := range payload.Acks {
+		if len(ack)+currentSize+1 > maxSizeBytes {
+			// We check if it's valid as it might be that the initial message
+			// is too big, in this case we still batch it
+			if newPayload.IsValid() {
+				response = append(response, newPayload)
+			}
+			newPayload = &protobuf.Payload{Acks: [][]byte{ack}}
+			currentSize = len(ack) + payloadTagSize + 1
+		} else {
+			newPayload.Acks = append(newPayload.Acks, ack)
+			currentSize += len(ack)
+		}
 	}
 
-	return t.dispatch(context.TODO(), publicKey, data, &payload)
+	for _, offer := range payload.Offers {
+		if len(offer)+currentSize+1 > maxSizeBytes {
+			if newPayload.IsValid() {
+				response = append(response, newPayload)
+			}
+			newPayload = &protobuf.Payload{Offers: [][]byte{offer}}
+			currentSize = len(offer) + payloadTagSize + 1
+		} else {
+			newPayload.Offers = append(newPayload.Offers, offer)
+			currentSize += len(offer)
+		}
```

Comment:
Can't help to notice that this block is identical to the one for `payload.Requests` and `playload.Acks`. Couldn't we make a helper function and call if in all three `for` loops? The third one is a bit different so it might not be worth generalizing it, but at least these three loops could be. Or do you think it wouldn't make much of a difference?

If it has a good name it could make it more clear about what's happening in those loops.

Reply:
I have the feeling that it will be more complex, as we'd have to generalize offer different types/methods, which is something that golang probably does not do very well, so I'd be inclined to keep it as it is, dumb but simple, with some duplication

Reply:
That's fair, if a bit ugly.